### PR TITLE
[CreateLogicalObjectFifoLink] Add support for multi-consumer/producer

### PR DESCRIPTION
**Pull request by jtuyls**
This PR adds support for multiple producer or consumer DMAs operating on a single logical objectfifo but accessing different parts. See for example the following test case, where there are three producer DMAs (`%3`, `%4`, `%5`) into a logical objectFifo on L2 (`%1`), but each DMA is writing into a different part (1: [0, 1024[, 2: [1024, 2048[, 3: [2048, 3072[). On the AIE dialect side this is implicitly specified through the order in the link operation (for now), so therefore you will see a sort on the base offset in the transformation and offsets being removed.

```
func.func @link_multiple_inputs_with_offsets(%arg0: memref<32x1024xi32>, %arg1: memref<3x32x64xi32, 1>, %arg2: memref<8x8x4x8xi32, 2>) {
  %0 = amdaie.logicalobjectfifo.from_memref %arg0, {} :memref<32x1024xi32> -> !amdaie.logicalobjectfifo<memref<32x1024xi32>>
  %1 = amdaie.logicalobjectfifo.from_memref %arg1, {} : memref<3x32x64xi32, 1> -> !amdaie.logicalobjectfifo<memref<3x32x64xi32, 1>>
  %2 = amdaie.logicalobjectfifo.from_memref %arg2, {} : memref<8x8x4x8xi32, 2> -> !amdaie.logicalobjectfifo<memref<8x8x4x8xi32, 2>>
  %3 = amdaie.circular_dma_cpy_nd(%1[0] [1024] [1], %0[] [] []) : (!amdaie.logicalobjectfifo<memref<3x32x64xi32, 1>>, !amdaie.logicalobjectfifo<memref<32x1024xi32>>)
  %4 = amdaie.circular_dma_cpy_nd(%1[1, 0] [1, 1024] [2048, 1], %0[] [] []) : (!amdaie.logicalobjectfifo<memref<3x32x64xi32, 1>>, !amdaie.logicalobjectfifo<memref<32x1024xi32>>)
  %5 = amdaie.circular_dma_cpy_nd(%1[1, 0] [1, 1024] [1024, 1], %0[] [] []) : (!amdaie.logicalobjectfifo<memref<3x32x64xi32, 1>>, !amdaie.logicalobjectfifo<memref<32x1024xi32>>)
  %6 = amdaie.circular_dma_cpy_nd(%2[] [] [], %1[] [] []) : (!amdaie.logicalobjectfifo<memref<8x8x4x8xi32, 2>>, !amdaie.logicalobjectfifo<memref<3x32x64xi32, 1>>)
  return
}
```

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/create_logical_objectfifo_link.mlir [hunk 1]
```diff
@@ -37,6 +37,40 @@ func.func @link_multiple_inputs(%arg0: memref<32x1024xi32>, %arg1: memref<32x64x
 
 // -----
 
+// CHECK-LABEL: func.func @link_multiple_inputs_with_offsets
+// CHECK:       %[[DMA0:.+]] = amdaie.circular_dma_cpy_nd
+// CHECK:       %[[DMA1:.+]] = amdaie.circular_dma_cpy_nd
+// CHECK:       %[[DMA2:.+]] = amdaie.circular_dma_cpy_nd
+// CHECK:       %[[DMA3:.+]] = amdaie.circular_dma_cpy_nd
+// CHECK:       amdaie.logicalobjectfifo.link[%[[DMA0]], %[[DMA2]], %[[DMA1]]] -> [%[[DMA3]]] ()
```
Comment (newling) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/create_logical_objectfifo_link.mlir:45]: Add a comment explaining why you expect DMA2 to appear before DMA1 (because it has a lower overall offset).

**Review by newling — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp [hunk 2]
```diff
@@ -6,13 +6,111 @@
...
+    offsets = op.getTargetMixedOffsets();
+    strides = op.getTargetMixedStrides();
+  } else {
+    return std::nullopt;
+  }
+  for (auto &&[offset, stride] : llvm::zip(offsets, strides)) {
+    std::optional<int64_t> constantOffset = getConstantIntValue(offset);
+    std::optional<int64_t> constantStride = getConstantIntValue(stride);
+    // If offset is zero, we can just continue to the next one.
+    if (constantOffset && constantOffset.value() == 0) continue;
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp:38]: This way it handles static offset zero with dynamic stride as well.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp [hunk 3]
```diff
@@ -6,13 +6,111 @@
...
+  SmallVector<OpFoldResult> offsets;
+  SmallVector<OpFoldResult> strides;
+  if constexpr (OperateOn == CopyOpOperateOn::Source) {
+    offsets = op.getSourceMixedOffsets();
+    strides = op.getSourceMixedStrides();
+  } else if constexpr (OperateOn == CopyOpOperateOn::Target) {
+    offsets = op.getTargetMixedOffsets();
+    strides = op.getTargetMixedStrides();
+  } else {
+    return std::nullopt;
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp:32]: Good idea

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp [hunk 4]
```diff
@@ -6,13 +6,111 @@
...
+    } else {
+      return std::nullopt;
+    }
+  }
+  return baseOffset;
+}
+
+/// Utility to compute the static extent on either the source or target side of
+/// a doubly strided operation. The extent is the number of elements covered by
+/// the strided access pattern, starting from the base offset.
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp:50]: > Worked example for access pattern: (offsets: [3, 7], sizes: [2, 2], strides: [4, 2])
>
> base_index = (3 + 0)*4 + (7 + 0)*2
> max_index = (3 + 1)*4 + (7 + 1)*2
>
> max_index - base_index = 4+2.
>
> It seems though, with your implementation, you add an additional 2 (inner most stride) to this and I'm just struggling > to find an interpretation of what this is...

Leaving the offsets aside as they are cancel out, the access pattern of `sizes: [2, 2], strides: [4, 2]` is:

```
1 - 1 -
1 - 1 -
```

So, while the last index is 6, the range from the first element to the last element accessed contains 7 elements. The current logic would return extent 8 to represent the range of elements 'covered' by the access pattern (which is safe to use in this case as it's an upper bound), but adding 1 seems more accurate as I realize now you can make a case for 7 elements as the 8th one isn't accessed.

Now, usually we will see last stride == 1, so here is another access pattern example for `sizes: [2, 2], strides: [4, 1]`:

```
1 1
- -
1 1
```

For this, `product_{dim} {(size[dim] - 1)* stride[dim]} == (2 - 1) * 4 + (2 - 1) * 1 == 5`, but the range contains 6 elements.

> Also, some lit tests for correct errors would be helpful.

These tests are there: https://github.com/nod-ai/iree-amd-aie/pull/457/files#diff-fbe4aa2d257e589cf0e40d2bfc290847423148566a99b00e6f313d66e0a71da7R66

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp [hunk 5]
```diff
@@ -6,13 +6,111 @@
...
+  } else {
+    return std::nullopt;
+  }
+
+  for (auto &&[offset, size, stride] : llvm::zip(offsets, sizes, strides)) {
+    std::optional<int64_t> constantOffset = getConstantIntValue(offset);
+    std::optional<int64_t> constantSize = getConstantIntValue(size);
+    std::optional<int64_t> constantStride = getConstantIntValue(stride);
+    // If offset is non-zero, we can just continue to the next one.
+    if (constantOffset && constantOffset.value() != 0) continue;
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp:74]: This is a mistake, thanks for pointing out.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp [hunk 6]
```diff
@@ -13,6 +14,49 @@
 
 namespace mlir::iree_compiler::AMDAIE {
 
+/// Utility to verify that there are no overlapping access patterns between
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp:17]: > Just to check: you're not interested in whether the patterns actually access the same elements?

Correct, this would still be considered an overlap as they access the same region of the buffer and this is currently not handled by the backend.

> Assuming that this has nothing to do with interleaving/non-interleaving, what is the problem specifically that MLIR-AIE has? I'm coming at this with a very shallow understanding of mlir-aie op semantics..

It can be supported in HW, even when consumers/producers are accessing the same elements, however, the aie.objectfifo ops don't have semantics for this at the moment as the order and sizes are used to implicitly split a buffer access across multiple producer/consumer actors, see https://github.com/Xilinx/mlir-aie/blob/061bc4a01d97f58ff0f2d63ecef099ca9c41b8ee/include/aie/Dialect/AIE/IR/AIEOps.td#L1685. And this works, because in most use cases, you won't see this overlap happening, but we might want to support it at some point (and there is some work on supporting offsets in the link operation in MLIR-AIE).

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp:50]: Yes, this definition makes sense.

**Comment by jtuyls** [comment]
@newling All your comments should have been addressed, could you please check again?

To better test the extent calculation I moved it to a separate Utils directory with dedicated cpp tests: https://github.com/nod-ai/iree-amd-aie/pull/457/files#diff-455624ff472424c7ec58bd67e3439058611e49d5723586619948b411b4ff0103R36

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/create_logical_objectfifo_link.mlir (hunk 1 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/create_logical_objectfifo_link.mlir:48]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp [hunk 7]
```diff
@@ -6,13 +6,111 @@
...
+      extent += ((constantSize.value() - 1) * constantStride.value());
+    } else {
+      return std::nullopt;
+    }
+  }
+  // Add last stride again to get the extent, not the last access index. For
+  // example, last access index 1023 would return an extent of 1024. NOTE: the
+  // stride should be constant by now.
+  if (!strides.empty())
+    extent += getConstantIntValue(strides[strides.size() - 1]).value();
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp]: No applicable anymore as `1` is added instead

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp [hunk 8]
```diff
@@ -13,6 +14,49 @@
...
+/// Access pattern 2: (offsets: [1, 0, 0], sizes: [1, 2, 8], strides: [32, 32,
+/// 1])
+///
+/// In this case, access pattern 1 access elements in range [0, 63], while
+/// access pattern 2 access elements in range [32, 95]. Therefore,
+/// these two access patterns overlap by both accessing elements in range [32,
+/// 63].
+template <CopyOpOperateOn OperateOn>
+LogicalResult checkForNoOverlappingAccessPatterns(
+    SmallVector<std::pair<DoublyStridedCopyOpInterface, int64_t>> &stridedOps) {
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateLogicalObjectFifoLink.cpp]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp [hunk 9]
```diff
@@ -6,13 +6,111 @@
...
+    // Subtract 1 from the stride to no t overshoot
+    if (constantOffset && constantSize && constantStride) {
+      extent += ((constantSize.value() - 1) * constantStride.value());
+    } else {
+      return std::nullopt;
+    }
+  }
+  // Add last stride again to get the extent, not the last access index. For
+  // example, last access index 1023 would return an extent of 1024. NOTE: the
+  // stride should be constant by now.
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEDmaOpInterface.cpp]: Updated

**Review by newling — COMMENTED** [review]
Thanks Jorn, looks good, I like to google tests. I've just left one final comment about edge cases for extent.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Utils/Utils.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+#ifndef IREE_COMPILER_AMDAIE_UTILS_DMA_UTILS_H_
+#define IREE_COMPILER_AMDAIE_UTILS_DMA_UTILS_H_
+
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/IR/OpDefinition.h"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+/// Utility to compute the extent of the range of values by the provided size
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Utils/Utils.h:15]: This is not about tensors, but about an access pattern on a buffer/tensor. An empty access pattern actually means a contiguous access of all elements, for which I think zero is a reasonable return value as zero doesn't have another meaning on an access pattern (you're never going to use a DMA to read/write zero values). Similarly, in AIE buffer descriptors a wrap of 0 means a contiguous access of all remaining elements.

[the remark this answers is not in the document]
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Utils/Utils.h:15]: Ok. I'm still thinking of this as just a geometry problem. I'm fine with it as is, thanks for all your feedback

**Review by newling — APPROVED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Utils/Utils.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Utils/Utils.h:15]: I realized though I didn't document this properly, so I am still going to do that
